src/vm_cpu.rs

```diff
@@ -96,6 +100,246 @@ pub struct VmCpuTrapState {
+                    return Err(SbiError::InvalidParam);
+                }
+            }
+            Ok(Self {
+                counter_index,
+                counter_mask,
+                mask_index: 0,
+                num_counters,
+            })
+        } else {
+            Err(SbiError::InvalidParam)
+        }
+    }
+}
+
+impl Iterator for CounterMaskIter {
+    type Item = usize;
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if self.counter_mask == 0 || (self.counter_index + self.mask_index) == self.num_counters
+            {
+                break None;
+            }
+            let result = self.counter_index + self.mask_index;
+            let mask_bit_set = self.counter_mask & (1 << self.mask_index);
+            self.counter_mask &= !(1 << self.mask_index);
+            self.mask_index += 1;
+            if mask_bit_set != 0 {
+                break Some(result as usize);
+            }
+        }
+    }
+}
+
+impl VmPmuState {
+    /// Updates internal state for PMU counters.
+    /// This should be called following a successful SBI call to configure counters.
+    pub fn update_configured_counter(
+        &mut self,
+        platform_counter_index: PlatformCounterIndex,
+        config_flags: PmuCounterConfigFlags,
+        event_type: PmuEventType,
+        event_data: u64,
+    ) -> SbiResult<()> {
+        use PmuCounterState::*;
+        let new_state = CounterState {
+            config_flags,
+            event_type,
+            event_data,
+            value: 0,
+        };
+        let pmu_info = pmu::PmuInfo::get()?;
+        let counter_index = pmu_info.get_logical_counter_index(platform_counter_index)?;
+        let state = &mut self.counter_state[counter_index.0 as usize];
+        match state {
+            // If skip_match is set, counter must already be configured
+            Configured(c) | Started(c) if config_flags.get_skip_match() => {
+                if config_flags.get_auto_start() {
+                    *state = Started(*c);
+                }
+                Ok(())
+            }
+            NotConfigured if !config_flags.get_skip_match() => {
+                *state = if config_flags.get_auto_start() {
+                    Started(new_state)
+                } else {
+                    Configured(new_state)
+                };
+                Ok(())
+            }
+            _ => Err(SbiError::InvalidParam),
+        }
+    }
+
+    /// Updates internal state for PMU counters.
+    /// This should be called following a successful SBI call to start counters.
+    pub fn update_started_counters(
+        &mut self,
+        counter_index: LogicalCounterIndex,
+        counter_mask: LogicalCounterMask,
+    ) {
+        use PmuCounterState::*;
+        // Unwrap ok: counter_index and counter_mask were previously validated
+        let bitmask_iter = CounterMaskIter::new(counter_index.0, counter_mask.0).unwrap();
```

Comment:
This contract that index and mask must be valid isn't documented. The caller of this function might have guaranteed that, but do we need to bake in that assumption within the callee? i.e. can we return an error here and `unwrap()` in the caller instead, where it's more obvious from the context that the necessary validation has been performed?